src/wbase/Task.h

```diff
@@ -227,11 +227,14 @@ class Task : public util::CommandForThreadPool {
     bool getFragmentHasSubchunks() const { return _fragmentHasSubchunks; }
     int getSubchunkId() const { return _subchunkId; }
 
-    /// Do not alter the returned value. &&& ??? just return a copy?
+    /// Do not alter the returned value, returns a set of database tables.
+    /// Note: Copying would be safer, but it could be an expensive thing to copy.
+    ///       This only used by ChunkResourceRequest acquire. The set is usually
+    ///       short, but this is called fairly frequently.
     DbTableSet& getDbTbls() { return _dbTbls; }
 
-    /// Do not alter the returned value. &&& ??? just return a copy?
-    IntVector& getSubchunksVect() { return _subchunksVect; }
+    /// Returns a copy of the list of subchunk ids.
+    IntVector getSubchunksVect() { return _subchunksVect; }
```

Comment:
What's the point in returning a copy vs a `const` reference to the vector? Is there any concern with the life expectancy of the vector stored withing `Task`?

Reply:
Copying a vector of int is always fairly cheap, so no reason not to be safe.

Reply:
I disagree with the statement that:
```
Copying a vector of int is always fairly cheap, ...
```
It depends on a size of a vector. This requires 1) calling (executing) a copy c-tor, 2) allocating memory for the vector itself, 3) allocating a block of memory where the vector elements will be stored. This operation will be more expensive than returning a reference to the vector (64-bit address). The latter won't require any memory allocation or copy constructor, etc.

As far as my other comment (regarding safety) was concerned then what you say here:
```
...so no reason not to be safe.
```
is not what I was asking for. My question was about if there would be thread safety or object lifecycle concerns causing this method not to return a reference to an internal data member (the vector in this case), especially if this member is mutable. In this case, a copy of the vector would need to be returned under the protection of a lock on a mutex. Something like this:
```
class Task {
public:
    void modifySubchunksVect() {
        std::lock_guard<std::mutex> const lock(_mtx);
        // Now it's safe to modify _subchunksVect
    }
    IntVector getSubchunksVect() {
        std::lock_guard<std::mutex> const lock(_mtx);
        // Now it's safe to call a copy c-tor of the vector to make a copy
        return _subchunksVect;
    }
```
What's shown above is one of two cases when you want to return a copy of a vector rather than a reference to the vector. The second case when you don't want to return the reference would be if the scope (the Task) would get destroyed while you had a reference in the caller's context:
```
Task* task = new Task(...);
Task::IntVector const& vec = task->getSubchunksVect();
delete task;
// Now you can't use the reference to the vector
```

In the end, your options here would (depending on a specific thread-safety or life expectancy requirements) be either of those:
```
    // Absolutely safe for threads and life expectancy.
   / / Very inefficient
    IntVector getSubchunksVect() {
        std::lock_guard<std::mutex> const lock(_mtx);
        // Now it's safe to call a copy c-tor of the vector to make a copy
        return _subchunksVect;
    }

    // Unsafe for threads and life expectancy, or vector mutability.
    // Extremely efficient.
    IntVector const& getSubchunksVect() { eturn _subchunksVect; }
```
What you have in your code is neither safe nor efficient.

Reply:
You don't seem to be looking at the most recent commit, which is in the pull request. The current code doesn't match what it is in your comments. The reason it is safe is that both are set only in the constructor, making reading safe without locking at any point thereafter. Defining them as constants would be nice, but a bit tricky to actually get them defined. The only thing to worry about is the caller changing the contents of the data structure. I can add a comment to the definitions (my preference) or add locks and make copies, which are then copied again. Please note, these do get called fairly frequently.


    /// Returns a reference to _dbTbls.
    const DbTableSet& getDbTbls() const { return _dbTbls; }

    /// Return a reference to the list of subchunk ids.
    const IntVector& getSubchunksVect() const { return _subchunksVect; }